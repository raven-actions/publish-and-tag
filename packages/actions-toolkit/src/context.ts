import fs from 'fs'

export interface WebhookEvent {
  id: string
  name: string
  payload: WebhookPayloadWithRepository
  protocol?: 'http' | 'https'
  host?: string
  url?: string
}

export interface PayloadRepository {
  [key: string]: any
  full_name?: string
  name: string
  owner: {
    [key: string]: any
    login: string
    name?: string
  }
  html_url?: string
}

export interface WebhookPayloadWithRepository {
  [key: string]: any
  repository?: PayloadRepository
  issue?: {
    [key: string]: any
    number: number
    html_url?: string
    body?: string
  }
  pull_request?: {
    [key: string]: any
    number: number
    html_url?: string
    body?: string
  }
  sender?: {
    [key: string]: any
    type: string
  }
  action?: string
  installation?: {
    id: number
    [key: string]: any
  }
}

export class Context {
  /**
   * Webhook payload object that triggered the workflow
   */
  public payload: WebhookPayloadWithRepository
  /**
   * Name of the event that triggered the workflow
   */
  public event: string
  public sha: string
  public ref: string
  public workflow: string
  public action: string
  public actor: string

  constructor() {
    this.payload = process.env.GITHUB_EVENT_PATH ? JSON.parse(fs.readFileSync(process.env.GITHUB_EVENT_PATH, 'utf8')) : {}
    this.event = process.env.GITHUB_EVENT_NAME as string
    this.sha = process.env.GITHUB_SHA as string
    this.ref = process.env.GITHUB_REF as string
    this.workflow = process.env.GITHUB_WORKFLOW as string
    this.action = process.env.GITHUB_ACTION as string
    this.actor = process.env.GITHUB_ACTOR as string
  }

  public get issue() {
    const payload = this.payload

    let issue_number: number
    if (payload.issue) {
      // If it's an issue
      issue_number = payload.issue.number
    } else if (payload.pull_request) {
      // If it's a PR
      issue_number = payload.pull_request.number
    } else if (payload.number) {
      // Just sittin' there on the payload
      issue_number = payload.number
    } else {
      throw new Error('tools.context.issue cannot be used with this event, there is no issue or pull_request object.')
    }

    return {
      ...this.repo,
      issue_number
    }
  }

  public get pullRequest() {
    const payload = this.payload

    let pull_number: number
    if (payload.pull_request) {
      // If it's a PR, the API expects pull_number
      pull_number = payload.pull_request.number
    } else {
      throw new Error('tools.context.pullRequest cannot be used with this event, there is no pull_request object.')
    }

    return {
      ...this.repo,
      pull_number
    }
  }

  public get repo() {
    if (process.env.GITHUB_REPOSITORY) {
      const [owner, repo] = process.env.GITHUB_REPOSITORY.split('/')
      return {owner, repo}
    }

    if (this.payload.repository) {
      return {
        owner: this.payload.repository.owner.login,
        repo: this.payload.repository.name
      }
    }

    throw new Error("context.repo requires a GITHUB_REPOSITORY environment variable like 'owner/repo'")
  }
}
