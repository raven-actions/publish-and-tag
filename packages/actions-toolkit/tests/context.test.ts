import {Context} from '../src/context.js'

describe('Context', () => {
  let context: Context

  beforeEach(() => {
    context = new Context()
  })

  describe('.payload', () => {
    it('returns the payload object', () => {
      expect(context.payload).toMatchSnapshot()
    })

    it('returns an empty object if the GITHUB_EVENT_PATH environment variable is falsey', () => {
      // Have to store the env var to pass later tests
      const before = process.env.GITHUB_EVENT_PATH
      delete process.env.GITHUB_EVENT_PATH

      context = new Context()
      expect(context.payload).toEqual({})

      // Reset it
      process.env.GITHUB_EVENT_PATH = before
    })
  })

  describe('#repo', () => {
    it('returns attributes from the GITHUB_REPOSITORY', () => {
      expect(context.repo).toEqual({owner: 'JasonEtco', repo: 'test'})
    })

    it('returns attributes from the repository payload', () => {
      context.payload.repository = {
        name: 'you',
        owner: {login: 'github'}
      }

      const before = process.env.GITHUB_REPOSITORY
      delete process.env.GITHUB_REPOSITORY
      expect(context.repo).toEqual({owner: 'github', repo: 'you'})
      process.env.GITHUB_REPOSITORY = before
    })

    it("return error for context.repo when repository doesn't exist", () => {
      delete context.payload.repository
      const before = process.env.GITHUB_REPOSITORY
      delete process.env.GITHUB_REPOSITORY
      expect(() => context.repo).toThrowErrorMatchingSnapshot()
      process.env.GITHUB_REPOSITORY = before
    })
  })

  describe('#issue', () => {
    it('returns attributes from the repository payload', () => {
      expect(context.issue).toEqual({owner: 'JasonEtco', repo: 'test', issue_number: 1})
    })

    it('works with pull_request payloads', () => {
      context.payload = {
        pull_request: {number: 2},
        repository: {owner: {login: 'JasonEtco'}, name: 'test'}
      }
      expect(context.issue).toEqual({
        issue_number: 2,
        owner: 'JasonEtco',
        repo: 'test'
      })
    })

    it('works with payload.number payloads', () => {
      context.payload = {number: 2, repository: {owner: {login: 'JasonEtco'}, name: 'test'}}
      expect(context.issue).toEqual({
        issue_number: 2,
        owner: 'JasonEtco',
        repo: 'test'
      })
    })

    it('throws if no number was found', () => {
      context.payload = {
        repository: {owner: {login: 'JasonEtco'}, name: 'test'}
      }

      expect(() => context.issue).toThrow()
    })
  })

  describe('#pullRequest', () => {
    it('returns attributes from the repository payload', () => {
      context.payload = {
        pull_request: {number: 2},
        repository: {owner: {login: 'JasonEtco'}, name: 'test'}
      }

      expect(context.pullRequest).toEqual({
        owner: 'JasonEtco',
        repo: 'test',
        pull_number: 2
      })
    })

    it('throws if no pull_request object was found', () => {
      context.payload = {
        repository: {owner: {login: 'JasonEtco'}, name: 'test'}
      }

      expect(() => context.pullRequest).toThrow()
    })
  })
})
