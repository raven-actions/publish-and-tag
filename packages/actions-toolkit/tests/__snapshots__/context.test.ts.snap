// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context #repo return error for context.repo when repository doesn't exist 1`] = `"context.repo requires a GITHUB_REPOSITORY environment variable like 'owner/repo'"`;

exports[`Context .payload returns the payload object 1`] = `
Object {
  "action": "opened",
  "issue": Object {
    "number": 1,
  },
  "repository": Object {
    "name": "test",
    "owner": Object {
      "login": "JasonEtco",
    },
  },
  "sender": Object {
    "type": "User",
  },
}
`;
